ratecounter.go

```diff
@@ -40,6 +41,12 @@ func (r *RateCounter) WithResolution(resolution int) *RateCounter {
 	return r
 }
 
+// OnStop allow to specify a function that will be called when the counter reaches 0
+// useful for removing it
+func (r *RateCounter) OnStop(f func(*RateCounter)) {
+	r.onStop = f
```

Comment:
To be thread-safe this either needs a mutex (which we would have to claim each time we reach zero), or to use an atomic pointer. The atomic pointer would be faster than locking/unlocking the mutex, but we would have to cast to and from a function which is a bit messy. The mutex will be slightly slower, but more type safety, and we only need to lock it when we reach zero anyway, so...
Whichever you feel comfortable implementing seems fine to me.